cdap-tms/src/main/java/co/cask/cdap/messaging/distributed/LeaderElectionMessagingService.java

```diff
@@ -195,16 +205,46 @@ private HBaseTableFactory getHBaseTableFactory(TableFactory tableFactory) {
     return factory instanceof HBaseTableFactory ? (HBaseTableFactory) factory : null;
   }
 
+  /**
+   * Updates the delegate with the given {@link DelegateService} and stop the old one.
+   * It also mark the delegate not usable.
+   */
   private void updateDelegate(@Nullable DelegateService newService) {
-    DelegateService oldService = delegate.getAndSet(newService);
+    DelegateService oldService = delegate.getReference();
+    while (!delegate.compareAndSet(oldService, newService, delegate.isMarked(), false)) {
+      oldService = delegate.getReference();
+    }
+
     if (oldService != null) {
       oldService.stopAndWait();
     }
   }
 
+  private void fencingStart(final DelegateService service) {
+    Runnable runnable = new Runnable() {
+      @Override
+      public void run() {
+        service.startAndWait();
+        // If failed to mark the service to become available, this means the follower() call happened before this,
+        // so just go ahead and shutdown the service.
+        if (!delegate.attemptMark(service, true)) {
```

Comment:
javadoc seems to indicate that the attemptMark method might fail spuriously. Would that affect the behavior?